### PR TITLE
tcp support and retries and timeout

### PR DESCRIPTION
**Pull request by Saif-Ali-Kamal**

### CODE REVIEW

File: src/components/deployments/routing-modal/RoutingModal.jsx [hunk 1]
```diff
@@ -17,8 +18,18 @@ const RoutingRule = props => {
...
+      if(values.protocol === "http"){
+        values.requestRetries = Number(values.requestRetries);
+        values.requestTimeout = Number(values.requestTimeout); 
+      }
+      let id = '';
+      if(mode === 'add'){
+        id = generateId();
+      }else{
+        id = initialValues.id;
+      }
```
Comment (YourTechBud) [inline src/components/deployments/routing-modal/RoutingModal.jsx:31]: This works right?

File: src/components/deployments/routing-modal/RoutingModal.jsx [hunk 2]
```diff
@@ -34,27 +45,52 @@ const RoutingRule = props => {
...
+            <Col>
+              <Form.Item name="port" rules={[
+                {
+                  validator: (_, value, cb) => {
+                    if (!value) {
+                      cb("Please provide a port value!")
+                      return
+                    }
+                    if (!Number.isInteger(Number(value))) {
+                      cb("Not a valid port value")
```
Comment (YourTechBud) [inline src/components/deployments/routing-modal/RoutingModal.jsx:73]:
```suggestion
                      cb("Port number should be a valid Integer")
```

File: src/mirage/fixtures.js [hunk 3]
```diff
@@ -749,8 +749,11 @@ export const serviceRoutes = [
   {
     "id": "service1",
     "source": {
+      "protocol": "tcp",
```
Comment (YourTechBud) [inline src/mirage/fixtures.js:752]:
```suggestion
      "protocol": "http",
```

File: src/mirage/fixtures.js [hunk 4]
```diff
@@ -749,8 +749,11 @@ export const serviceRoutes = [
...
       "port": 8080
     },
+    "requestRetries": 7,
+    "requestTimeout": 450,
```
Comment (YourTechBud) [inline src/mirage/fixtures.js:756]:
```suggestion
    "requestRetries": 3,
    "requestTimeout": 180,
```

File: src/mirage/fixtures.js [hunk 5]
```diff
@@ -749,8 +749,11 @@ export const serviceRoutes = [
   {
```
Comment (YourTechBud) [inline src/mirage/fixtures.js:749]: A suggest you add one more tcp route in here

File: src/pages/deployments/routes/DeploymentsRoutes.jsx [hunk 6]
```diff
@@ -39,6 +39,13 @@ const DeploymentsRoutes = () => {
     if (!serviceRoutes[obj.id]) {
       serviceRoutes[obj.id] = []
     }
+
+    serviceRoutes[obj.id].forEach(route => {
+      if(!route.id) route.id = generateId();
```
Comment (YourTechBud) [inline src/pages/deployments/routes/DeploymentsRoutes.jsx:44]:
```suggestion
      if (!route.id) route.id = generateId();
```

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: src/components/deployments/routing-modal/RoutingModal.jsx (hunk 1 above)
Comment (Saif-Ali-Kamal) [reply src/components/deployments/routing-modal/RoutingModal.jsx:31]: yes it works

File: src/mirage/fixtures.js (hunk 5 above)
Comment (Saif-Ali-Kamal) [reply src/mirage/fixtures.js:749]: done

File: src/operations/deployments.js [hunk 7]
```diff
@@ -176,14 +176,21 @@ const saveServiceRoutesConfig = (projectId, serviceId, serviceRoutes) => {
 export const saveServiceRoutes = (projectId, serviceId, routeConfig) => {
   const serviceRoutes = getServiceRoutes(store.getState())
   const serviceRoute = get(serviceRoutes, serviceId, [])
-  const newServiceRoutes = [...serviceRoute.filter(obj => obj.source.port !== routeConfig.source.port), routeConfig]
+  let newServiceRoutes = [];
```
Comment (YourTechBud) [inline src/operations/deployments.js:179]: You are doing a map. No need to follow this approach

**Review by YourTechBud — CHANGES_REQUESTED** [review]

**Review by YourTechBud — APPROVED** [review]

**Pull request merged**